build.gradle

```diff
@@ -582,3 +514,56 @@ subprojects {
         }
     }
 }
+
+class DepAndParents {
+    DependencyResult dep
+    List<String> parents
+}
+
+/**
+ * Make sure that Maven would select the same versions as Gradle selected.
+ * This is essentially the same as if we used Maven Enforcer's
+ * requireUpperBoundDeps for our artifacts.
+ */
+def requireUpperBoundDepsMatch(Configuration conf, Project project) {
+    // artifact name => version
+    Map<String,String> golden = conf.resolvedConfiguration.resolvedArtifacts.collectEntries {
+        ResolvedArtifact it ->
+            ModuleVersionIdentifier id = it.moduleVersion.id
+            [id.group + ":" + id.name, id.version]
+    }
+    // Breadth-first search like Maven for dependency resolution
+    Queue<DepAndParents> queue = new ArrayDeque<>()
+    conf.incoming.resolutionResult.root.dependencies.each {
+        queue.add(new DepAndParents(dep: it, parents: ["" + project]))
+    }
+    Set<String> found = new HashSet<>()
+    while (!queue.isEmpty()) {
+        DepAndParents depAndParents = queue.remove()
+        ResolvedDependencyResult result = (ResolvedDependencyResult) depAndParents.dep
+        ModuleVersionIdentifier id = result.selected.moduleVersion
+        String artifact = id.group + ":" + id.name
+        if (found.contains(artifact))
+            continue
+        found.add(artifact)
+        String version
+        if (result.requested instanceof ProjectComponentSelector) {
+            ProjectComponentSelector selector = (ProjectComponentSelector) result.requested
+            version = project.findProject(selector.projectPath).version
+        } else {
+            version = ((ModuleComponentSelector) result.requested).version
+        }
+        String goldenVersion = golden[artifact]
+        if (goldenVersion != version && "[$goldenVersion]" != version) {
+            throw new RuntimeException(
+                "Maven version skew: $artifact ($version != $goldenVersion) "
+                + "Bad version dependency path: " + depAndParents.parents
+                + " Run './gradlew $project.path:dependencies --configuration $conf.name' "
```

Comment:
`./gradlew $project.path:dependencyInsight --dependency $artifact --configuration $conf.name` is probably more helpful.

Reply:
I hadn't considered that. But I tried it out and it produces worse results. The main problem is that the order of the dependencies matters for Maven but doesn't for Gradle. The dependencies task actually outputs the dependencies in their real order (as it would be in a pom). But the dependencyInsight is trying to show how _Gradle_ uses dependencies and those have no natural ordering and it produces more noise.

I introduced an old version of Guava to force a breakage, and compared the results. Dependencies:
```
runtimeClasspath - Runtime classpath of source set 'main'.
+--- project :grpc-api
|    +--- project :grpc-context
|    +--- com.google.code.findbugs:jsr305:3.0.2
|    \--- com.google.guava:guava:30.1-android
|         +--- com.google.guava:failureaccess:1.0.1
|         +--- com.google.guava:listenablefuture:9999.0-empty-to-avoid-conflict-with-guava
|         +--- com.google.code.findbugs:jsr305:3.0.2
|         +--- org.checkerframework:checker-compat-qual:2.5.5
|         +--- com.google.errorprone:error_prone_annotations:2.3.4 -> 2.4.0
|         \--- com.google.j2objc:j2objc-annotations:1.3
+--- com.google.guava:guava:29.0-android -> 30.1-android (*)
+--- com.google.code.gson:gson:2.8.6
+--- com.google.android:annotations:4.1.1.4
+--- org.codehaus.mojo:animal-sniffer-annotations:1.19
+--- com.google.errorprone:error_prone_annotations:2.4.0
+--- com.google.guava:guava:30.1-android (*)
\--- io.perfmark:perfmark-api:0.23.0
```

dependencyInsight:
```
com.google.guava:guava:30.1-android
   variant "runtime" [
      org.gradle.status              = release (not requested)
      org.gradle.usage               = java-runtime
      org.gradle.libraryelements     = jar
      org.gradle.category            = library

      Requested attributes not found in the selected variant:
         org.gradle.dependency.bundling = external
         org.gradle.jvm.version         = 7
   ]
   Selection reasons:
      - By conflict resolution : between versions 30.1-android and 29.0-android

com.google.guava:guava:30.1-android
+--- project :grpc-api
|    \--- runtimeClasspath
\--- runtimeClasspath

com.google.guava:guava:29.0-android -> 30.1-android
\--- runtimeClasspath
```

dependencyInsight could work, but showing it "like Maven" helps me the most.